collectstatic chokes when linked map files aren't in node_libs in django4.x

part of https://github.com/mozilla/addons-server/issues/19595 to support newer 4.x django - collectstatic now looks for linked files in js and css files, and (at least some of) the js dependencies include links to the xxx.js.map file.  

We don't currently copy those files over and there is no obvious command line switch to disable the feature.  (I tried `--ignore '*.map'` but didn't seem to make a difference)

> Post-processing 'js/node_lib/underscore.js' failed!
> 
> Traceback (most recent call last):
>   File "/home/circleci/project/manage.py", line 11, in <module>
>     execute_from_command_line(sys.argv)
>   File "/home/circleci/.pyenv/versions/3.10.6/lib/python3.10/site-packages/django/core/management/__init__.py", line 446, in execute_from_command_line
>     utility.execute()
>   File "/home/circleci/.pyenv/versions/3.10.6/lib/python3.10/site-packages/django/core/management/__init__.py", line 440, in execute
>     self.fetch_command(subcommand).run_from_argv(self.argv)
>   File "/home/circleci/.pyenv/versions/3.10.6/lib/python3.10/site-packages/django/core/management/base.py", line 402, in run_from_argv
>     self.execute(*args, **cmd_options)
>   File "/home/circleci/.pyenv/versions/3.10.6/lib/python3.10/site-packages/django/core/management/base.py", line 448, in execute
>     output = self.handle(*args, **options)
>   File "/home/circleci/.pyenv/versions/3.10.6/lib/python3.10/site-packages/django/contrib/staticfiles/management/commands/collectstatic.py", line 209, in handle
>     collected = self.collect()
>   File "/home/circleci/.pyenv/versions/3.10.6/lib/python3.10/site-packages/django/contrib/staticfiles/management/commands/collectstatic.py", line 154, in collect
>     raise processed
>   File "/home/circleci/.pyenv/versions/3.10.6/lib/python3.10/site-packages/django/contrib/staticfiles/storage.py", line 338, in _post_process
>     content = pattern.sub(converter, content)
>   File "/home/circleci/.pyenv/versions/3.10.6/lib/python3.10/site-packages/django/contrib/staticfiles/storage.py", line 215, in converter
>     hashed_url = self._url(
>   File "/home/circleci/.pyenv/versions/3.10.6/lib/python3.10/site-packages/django/contrib/staticfiles/storage.py", line 152, in _url
>     hashed_name = hashed_name_func(*args)
>   File "/home/circleci/.pyenv/versions/3.10.6/lib/python3.10/site-packages/django/contrib/staticfiles/storage.py", line 388, in _stored_name
>     cache_name = self.clean_name(self.hashed_name(name))
>   File "/home/circleci/.pyenv/versions/3.10.6/lib/python3.10/site-packages/django/contrib/staticfiles/storage.py", line 113, in hashed_name
>     raise ValueError(
> ValueError: The file 'js/node_lib/underscore-umd.js.map' could not be found with <django.contrib.staticfiles.storage.ManifestStaticFilesStorage object at 0x7f53454d4b50>.
> make: *** [Makefile-docker:123: update_assets] Error 1
> 
> Exited with code exit status 2
> 
> CircleCI received exit code 2
>

--- a/src/olympia/lib/settings_base.py
+++ b/src/olympia/lib/settings_base.py
@@ -1263,7 +1263,7 @@ STATIC_ROOT = path('site-static')
 STATIC_URL = '/static/'
 
 STATICFILES_DIRS = (path('static'),)
-STATICFILES_STORAGE = 'django.contrib.staticfiles.storage.ManifestStaticFilesStorage'
+STATICFILES_STORAGE = 'olympia.lib.storage.ManifestStaticFilesStorageNotMaps'
 
 # Path related settings. In dev/stage/prod `NETAPP_STORAGE_ROOT` environment
 # variable will be set and point to our NFS/EFS storage

--- a/src/olympia/lib/storage.py
+++ b/src/olympia/lib/storage.py
@@ -6,11 +6,14 @@ class ManifestStaticFilesStorageNotMaps(ManifestStaticFilesStorage):
         (
             '*.css',
             (
+                # These regexs are copied from HashedFilesMixin in django4.1+
                 r"""(?P<matched>url\(['"]{0,1}\s*(?P<url>.*?)["']{0,1}\))""",
                 (
                     r"""(?P<matched>@import\s*["']\s*(?P<url>.*?)["'])""",
                     """@import url("%(url)s")""",
                 ),
+                # We are ommiting the sourceMappingURL regexs for .css and .js as they
+                # don't work how we copy over the souces in Makefile-docker.copy_node_js
             ),
         ),
     )

--- a/src/olympia/lib/storage.py
+++ b/src/olympia/lib/storage.py
@@ -1,0 +1,16 @@
+from django.contrib.staticfiles.storage import ManifestStaticFilesStorage
+
+
+class ManifestStaticFilesStorageNotMaps(ManifestStaticFilesStorage):
+    patterns = (
+        (
+            '*.css',
+            (
+                r"""(?P<matched>url\(['"]{0,1}\s*(?P<url>.*?)["']{0,1}\))""",
+                (
+                    r"""(?P<matched>@import\s*["']\s*(?P<url>.*?)["'])""",
+                    """@import url("%(url)s")""",
+                ),
+            ),
+        ),
+    )

